internal/handler/ticket.go

```diff
@@ -2,23 +2,82 @@ package handler
 
 import (
 	"context"
+	"github.com/dilyara4949/flight-booking-api/internal/middleware"
+	"log/slog"
 	"net/http"
 
 	"github.com/dilyara4949/flight-booking-api/internal/domain"
 	"github.com/dilyara4949/flight-booking-api/internal/handler/request"
 	"github.com/dilyara4949/flight-booking-api/internal/handler/response"
-	"github.com/dilyara4949/flight-booking-api/internal/middleware"
 	"github.com/gin-gonic/gin"
 	"github.com/google/uuid"
 )
 
 type TicketService interface {
+	BookTicket(ctx context.Context, req request.BookTicket, userID uuid.UUID, flight domain.Flight) (domain.Ticket, error)
+	CheckAvailability(ctx context.Context, flightID uuid.UUID, totalTickets int) (bool, error)
 	Get(ctx context.Context, ticketID, userID uuid.UUID) (domain.Ticket, error)
 	Delete(ctx context.Context, ticketID, userID uuid.UUID) error
 	GetTickets(ctx context.Context, userID uuid.UUID, page, pageSize int) ([]domain.Ticket, error)
 	Update(ctx context.Context, ticketID, userID uuid.UUID, req request.UpdateTicket) (domain.Ticket, error)
 }
 
+func BookTicketHandler(ticketService TicketService, flightService FlightService) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if !AccessCheck(*c, c.GetString(middleware.UserIDKey), userIDParamKey) {
+			c.JSON(http.StatusForbidden, response.Error{Error: "access denied"})
+			return
+		}
+
+		var req request.BookTicket
+
+		if err := c.ShouldBind(&req); err != nil {
+			slog.Error("error at binding request body", "error", err.Error())
+
+			c.JSON(http.StatusBadRequest, response.Error{Error: "error at binding request body"})
+
+			return
+		}
+
+		if req.FlightID == uuid.Nil {
+			c.JSON(http.StatusBadRequest, response.Error{Error: response.ErrEmptyRequestFields.Error()})
+			return
+		}
+
+		userID, err := uuid.Parse(c.Param(userIDParamKey))
+		if err != nil {
+			c.JSON(http.StatusBadRequest, response.Error{Error: "id is not correct"})
+			return
+		}
+
+		flight, err := flightService.Get(c, req.FlightID, true)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, response.Error{Error: err.Error()})
+			return
+		}
+
+		available, err := ticketService.CheckAvailability(c, req.FlightID, flight.TotalTickets)
```

Comment:
In theory, I could see a potential racing error, such as many people booking their tickets faster than the check is performed, resulting in overbooking. I would consider checking availability and booking as a part of the same transactional action and/or trigger or constraint on table action to make sure that we're not overbooking (although it happens in a real world, but as part of design, not a bug). You can verify, that this behavior can occur with load testing.